### PR TITLE
[4.0] Normalize finder tables and add default values

### LINKED ISSUE #13509: Mssql: Adding a tag doesn´t work
### Steps to reproduce the issue
On saving a new tag you get:
515 [Microsoft][SQL Server Native Client 11.0][SQL Server]Cannot insert the value NULL into column 'term_id', table 'joomla35.dbo.#__finder_tokens_aggregate'; column does not allow nulls. INSERT fails.

### PR DESCRIPTION
**Pull request by Hackwar**
This is a new version of #13517 which originally was supposed to fix #13509. Since all of my finder work is aimed at 4.0, this PR is, too. I also extended this further and since MSSQL is not supported in 4.0 anymore, I've not adopted those changes.
Since this PR is now open for nearly a year, I also reworked this PR to incorporate all DB changes to the finder tables in one big update file instead of several smaller ones.

## Changes
* Changed all field types to be identical across DB systems and across tables
* Added default values for all relevant fields
* Added indexes where appropriate
* Updated component SQL files
* Merged all DB changes for 4.0 for Finder into one update file

Considering all the changes that happened to finder, a complete re-indexing is necessary. This requirement is not related to this PR, but comes from the removal of the sharding, the new taxonomies and several other changes. This also means that the filters will have to be manually recreated. All these updates can be done on filled tables, but that would make little sense, since we have to run a complete indexing run anyway, thus this PR first clears the tables and then does the updates. That should help with timeout issues for large indexes.

## How to test
I'm quoting the original dev of this PR:
1. SQL code review
2. Test update: Apply patch, manually run the sql query for your Db system, confirm db table, confirm all fine
3. Test install: apply patch, delete configuration.php do a normal install, confirm db table, confirm all fine

@alikon Can you check the Postgres code? Is there a difference between VARCHAR and CHAR in Postgres? Maybe we want to change the language columns from VARCHAR to CHAR? I'm confused because the original author did not change those.

### CODE REVIEW

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-21.sql [hunk 1]
```diff
@@ -0,0 +1,67 @@
+-- Normalize finder tables default values.
+-- finder_filters table
+ALTER TABLE `#__finder_filters` MODIFY `title` varchar(255) NOT NULL DEFAULT '';
+ALTER TABLE `#__finder_filters` MODIFY `alias` varchar(255) NOT NULL DEFAULT '';
+ALTER TABLE `#__finder_filters` MODIFY `created_by` int(10) unsigned NOT NULL DEFAULT 0;
+ALTER TABLE `#__finder_filters` MODIFY `created_by_alias` varchar(255) NOT NULL DEFAULT '';
+ALTER TABLE `#__finder_filters` MODIFY `data` text NOT NULL DEFAULT '';
```
Comment (csthomas) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-21.sql]: The above/below lines are wrong. 
The `text/mediumtext` type can not have a default value other than NULL. 

Mysql simply ignores this query. 

After adjusting the php code, correct should be 
`ALTER TABLE `#__finder_filters` MODIFY `data` text NULL DEFAULT NULL;` 
or simpler 
`ALTER TABLE `#__finder_filters` MODIFY `data` text;` 
but I'm not sure if joomla can undestand it.

**Comment by Hackwar** [comment]
With the changes in #20185 merged into this branch and conflicts fixed, this is ready to merge. :wink:

File: administrator/components/com_finder/sql/install.mysql.sql [hunk 2]
```diff
@@ -26,29 +26,30 @@ CREATE TABLE IF NOT EXISTS `#__finder_filters` (
 
 CREATE TABLE IF NOT EXISTS `#__finder_links` (
   `link_id` int(10) unsigned NOT NULL AUTO_INCREMENT,
-  `url` varchar(255) NOT NULL,
-  `route` varchar(255) NOT NULL,
+  `url` varchar(255) NOT NULL DEFAULT '',
+  `route` varchar(255) NOT NULL DEFAULT '',
   `title` varchar(400) DEFAULT NULL,
-  `description` varchar(255) DEFAULT NULL,
+  `description` text NOT NULL DEFAULT '',
```
Comment (csthomas) [inline administrator/components/com_finder/sql/install.mysql.sql]: It should be only "`description` text," in other places too.

**Comment by Hackwar** [comment]
Do we need to remove the default value for postgres, too? I couldn't find anything if the text field in postgres has a default value or not. SQL standards. A joy for everyone...

File: administrator/components/com_finder/sql/install.postgresql.sql [hunk 3]
```diff
@@ -1,53 +1,56 @@
...
--- Table: #__finder_links
+-- Table structure for table `#__finder_links`
 --
+
 CREATE TABLE IF NOT EXISTS "#__finder_links" (
   "link_id" serial NOT NULL,
-  "url" character varying(255) NOT NULL,
-  "route" character varying(255) NOT NULL,
-  "title" character varying(255) DEFAULT NULL,
-  "description" character varying(255) DEFAULT NULL,
```
Comment (csthomas) [inline administrator/components/com_finder/sql/install.postgresql.sql:30]: This line and the same in other postgresql files should/can be set as  `"description" text,`

**Comment by csthomas** [comment]
I found https://www.postgresql.org/docs/9.3/static/ddl-default.html

> If no default value is declared explicitly, the default value is the null value. 

IIRC default value can be set for text type in postgresql, 
but it would be good to have the same default value in both databases.

**Comment by Hackwar** [comment]
Updated the stuff according to your request

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-21.sql [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+-- finder_tokens table
+ALTER TABLE `#__finder_tokens` MODIFY `term` varchar(75) NOT NULL DEFAULT '';
+ALTER TABLE `#__finder_tokens` MODIFY `stem` varchar(75) NOT NULL DEFAULT '';
+ALTER TABLE `#__finder_tokens` MODIFY `weight` float unsigned NOT NULL DEFAULT 0;
+ALTER TABLE `#__finder_tokens` MODIFY `language` char(7) NOT NULL DEFAULT '';
+ALTER TABLE `#__finder_tokens` ADD INDEX `idx_stem` (`stem`);
+ALTER TABLE `#__finder_tokens` ADD INDEX `idx_language` (`language`);
+-- finder_tokens_aggregate table
+ALTER TABLE `#__finder_tokens_aggregate` MODIFY `term_id` int(10) unsigned NOT NULL DEFAULT 0;
+ALTER TABLE `#__finder_tokens_aggregate` MODIFY `map_suffix` char(1) NOT NULL DEFAULT '';
```
Comment (Quy) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-21.sql]: This column is dropped as of #20185

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-21.sql [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ALTER TABLE `#__finder_filters` MODIFY `created_by_alias` varchar(255) NOT NULL DEFAULT '';
+-- finder_links table
+ALTER TABLE `#__finder_links` MODIFY `url` varchar(255) NOT NULL DEFAULT '';
+ALTER TABLE `#__finder_links` MODIFY `route` varchar(255) NOT NULL DEFAULT '';
+ALTER TABLE `#__finder_links` MODIFY `state` int(5) NOT NULL DEFAULT 1;
+ALTER TABLE `#__finder_links` MODIFY `access` int(5) NOT NULL DEFAULT 0;
+ALTER TABLE `#__finder_links` MODIFY `language` char(7) NOT NULL DEFAULT '';
+ALTER TABLE `#__finder_links` MODIFY `type_id` int(11) NOT NULL DEFAULT 0;
+ALTER TABLE `#__finder_links` ADD INDEX `idx_language` (`language`);
+-- finder_links_terms tables
```
Comment (Quy) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-21.sql]: change to `table`

**Comment by Hackwar** [comment]
Fixed the comments from @Quy. This is now ready to test/review.

**Comment by brianteeman** [comment]
Why does this component have its own install.sql ??
Other components do not

**Comment by Hackwar** [comment]
These SQL files are (most likely) still from the origin of this component. Yes, we don't do this elsewhere. On the other hand, I actually like that we have this component in a state where you could just zip it up and install it into a site where it is missing. Maybe we should strive for that for our other components, too.

**Comment by TobsBobs** [comment]
This branch is out-of-date. Can I still test the sql changes or should I wait?

**Comment by Hackwar** [comment]
The out-of-date message can mostly be ignored, unless there is an additional message that there are conflicts. In any case I updated this PR to the latest changes.

**Comment by TobsBobs** [comment]
I have test the number 2 - update an installation!
I installed the update sql file and played a lot with the smart search. I could not detect any issues. Everything works as before.

When I figure out how I can put a patch into my system with git, then I'll test number 3 - new installation. ;)
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20561">issues.joomla.org/tracker/joomla-cms/20561</a>.</sub>

**Comment by csthomas** [comment]
If you have your own joomla on git then you can apply this patch by command like (on linux):
```bash
curl https://patch-diff.githubusercontent.com/raw/joomla/joomla-cms/pull/20561.diff | git apply
```

or simply download the patch from the above link and enter `git apply 20561.diff`

File: administrator/components/com_finder/sql/install.mysql.sql [hunk 6]
```diff
@@ -26,29 +26,30 @@ CREATE TABLE IF NOT EXISTS `#__finder_filters` (
 
 CREATE TABLE IF NOT EXISTS `#__finder_links` (
   `link_id` int(10) unsigned NOT NULL AUTO_INCREMENT,
-  `url` varchar(255) NOT NULL,
-  `route` varchar(255) NOT NULL,
+  `url` varchar(255) NOT NULL DEFAULT '',
+  `route` varchar(255) NOT NULL DEFAULT '',
```
Comment (wilsonge) [inline administrator/components/com_finder/sql/install.mysql.sql]: url and route being nullable seems weird to me

File: administrator/components/com_finder/sql/install.mysql.sql [hunk 7]
```diff
@@ -26,29 +26,30 @@ CREATE TABLE IF NOT EXISTS `#__finder_filters` (
...
+  `language` char(7) NOT NULL DEFAULT '',
   `publish_start_date` datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
   `publish_end_date` datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
   `start_date` datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
   `end_date` datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
   `list_price` double unsigned NOT NULL DEFAULT 0,
   `sale_price` double unsigned NOT NULL DEFAULT 0,
-  `type_id` int(11) NOT NULL,
-  `object` mediumblob NOT NULL,
+  `type_id` int(11) NOT NULL DEFAULT 0,
```
Comment (wilsonge) [inline administrator/components/com_finder/sql/install.mysql.sql]: I don't think this should default. If you're missing a type then things should blow up i think?

File: administrator/components/com_finder/sql/install.mysql.sql [hunk 8]
```diff
@@ -4,18 +4,18 @@
 
 CREATE TABLE IF NOT EXISTS `#__finder_filters` (
   `filter_id` int(10) unsigned NOT NULL AUTO_INCREMENT,
-  `title` varchar(255) NOT NULL,
-  `alias` varchar(255) NOT NULL,
+  `title` varchar(255) NOT NULL DEFAULT '',
+  `alias` varchar(255) NOT NULL DEFAULT '',
```
Comment (wilsonge) [inline administrator/components/com_finder/sql/install.mysql.sql]: should title and alias also be nullable? I'm not sure i can imagine a use case for this

File: administrator/components/com_finder/sql/install.mysql.sql [hunk 9]
```diff
@@ -124,28 +125,30 @@ CREATE TABLE IF NOT EXISTS `#__finder_taxonomy_map` (
 
 CREATE TABLE IF NOT EXISTS `#__finder_terms` (
   `term_id` int(10) unsigned NOT NULL AUTO_INCREMENT,
-  `term` varchar(75) NOT NULL,
-  `stem` varchar(75) NOT NULL,
+  `term` varchar(75) NOT NULL DEFAULT '',
```
Comment (wilsonge) [inline administrator/components/com_finder/sql/install.mysql.sql]: Ditto here in the terms table having an empty string term seems odd

File: administrator/components/com_finder/sql/install.mysql.sql [hunk 10]
```diff
@@ -124,28 +125,30 @@ CREATE TABLE IF NOT EXISTS `#__finder_taxonomy_map` (
...
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 DEFAULT COLLATE=utf8mb4_general_ci;
 
 --
 -- Table structure for table `#__finder_terms_common`
 --
 
 CREATE TABLE IF NOT EXISTS `#__finder_terms_common` (
-  `term` varchar(75) NOT NULL,
-  `language` char(7) NOT NULL,
+  `term` varchar(75) NOT NULL DEFAULT '',
```
Comment (wilsonge) [inline administrator/components/com_finder/sql/install.mysql.sql]: ditto here

**Comment by Hackwar** [comment]
I've adopted the requested changes. Please review this for merging. Thx.

**Comment by Hackwar** [comment]
I've rewritten this PR to also combine all DB changes for 4.0 of Finder into one file and consolidate the different changes. This one file should update all the table at once. I'm considering also adding truncate statements for all tables, since it is pretty useless to alter these tables, adding indexes and all that, when they are full of data and take ages to process, while they also all have to be deleted and re-indexed anyway.

**Comment by Hackwar** [comment]
I updated the initial description and did some last changes. Please advise how to progress with this.

File: installation/sql/mysql/joomla.sql [hunk 11]
```diff
@@ -1213,8 +1217,8 @@ CREATE TABLE IF NOT EXISTS `#__finder_tokens_aggregate` (
 
 CREATE TABLE IF NOT EXISTS `#__finder_types` (
   `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
-  `title` varchar(100) NOT NULL,
-  `mime` varchar(100) NOT NULL,
+  `title` varchar(100) NOT NULL DEFAULT '',
```
Comment (wilsonge) [inline installation/sql/mysql/joomla.sql]: Given title is a unique key, having a default here doesn't make sense. I think it needs to be compulsory

File: installation/sql/mysql/joomla.sql [hunk 12]
```diff
@@ -1173,15 +1175,17 @@ INSERT INTO `#__finder_terms_common` (`term`, `language`, `custom`) VALUES
 --
 
 CREATE TABLE IF NOT EXISTS `#__finder_tokens` (
-  `term` varchar(75) NOT NULL,
-  `stem` varchar(75) NOT NULL,
+  `term` varchar(75) NOT NULL DEFAULT '',
```
Comment (wilsonge) [inline installation/sql/mysql/joomla.sql]: Is it possible to have a token that doesn't have a term?

File: installation/sql/mysql/joomla.sql [hunk 13]
```diff
@@ -1173,15 +1175,17 @@ INSERT INTO `#__finder_terms_common` (`term`, `language`, `custom`) VALUES
...
+  `stem` varchar(75) NOT NULL DEFAULT '',
   `common` tinyint(1) unsigned NOT NULL DEFAULT 0,
   `phrase` tinyint(1) unsigned NOT NULL DEFAULT 0,
-  `weight` float unsigned NOT NULL DEFAULT 1,
+  `weight` float unsigned NOT NULL DEFAULT 0,
```
Comment (wilsonge) [inline installation/sql/mysql/joomla.sql]: What's the effect of changing default 0 to default 1?

File: installation/sql/mysql/joomla.sql [hunk 14]
```diff
@@ -1191,15 +1195,15 @@ CREATE TABLE IF NOT EXISTS `#__finder_tokens` (
 --
 
 CREATE TABLE IF NOT EXISTS `#__finder_tokens_aggregate` (
-  `term_id` int(10) unsigned NOT NULL,
-  `term` varchar(75) NOT NULL,
-  `stem` varchar(75) NOT NULL,
+  `term_id` int(10) unsigned NOT NULL DEFAULT 0,
```
Comment (wilsonge) [inline installation/sql/mysql/joomla.sql]: Given this is effectively a foreign key to the `#__finder_terms` i don't think a default here makes sense

File: installation/sql/mysql/joomla.sql [hunk 15]
```diff
@@ -1191,15 +1195,15 @@ CREATE TABLE IF NOT EXISTS `#__finder_tokens` (
...
+  `term` varchar(75) NOT NULL DEFAULT '',
```
Comment (wilsonge) [inline installation/sql/mysql/joomla.sql]: As a result to the above row, if `term_id` is compulsory, i guess the term is too

File: installation/sql/mysql/joomla.sql (hunk 13 above)
Comment (Hackwar) [reply installation/sql/mysql/joomla.sql]: I have no idea anymore why I did that. Changing it back.

**Comment by Hackwar** [comment]
I changed all the requested things...

**Comment by wilsonge** [comment]
LGTM. Thanks!

**Comment by richard67** [comment]
@Hackwar @wilsonge If this requires reindexing anyway, wouldn't it have been a good time to change the collation of the finder tables to utf8mb4_**unicode**_ci?

The only reason why we kept some finder tables with utf8mb4_general_ci was that we did not want to force a complete reindexing when doing the utf8mb4 conversion. But that mix of collations has caused some issues from time to time, e.g. when other 3rd oparty extensions doing joins to finder and other core tables.

If updating or migrating from 3.10 to 4.0 will anyway require a complete reindexing of finder, we can change collation in 4.0.

**Comment by Hackwar** [comment]
you are right. Would you create a PR to do that?

**Comment by wilsonge** [comment]
I thought we'd already done that :/ but apparently it got closed and not merged. requires resurrecting and careful testing of https://github.com/joomla/joomla-cms/pull/16617 but we should definitely do it

**Comment by richard67** [comment]
In principle it needs a redo of #16617 but without the conversion of existing data. Instead of this, tables should be cleared like in schema updates for this PR here.

Not sure if I will find the time for that soon.

**Comment by richard67** [comment]
It seems there is something wrong with this PR for PostgreSQL, see new PR #24771 for new issue #24769 .

**Pull request merged**